communication/nats/dialog/dialog_establisher.go

```diff
@@ -30,49 +30,82 @@ func NewDialogEstablisher(myIdentity identity.Identity, signer identity.Signer)
 }
 
-func (establisher *dialogEstablisher) CreateDialog(contact dto_discovery.Contact) (communication.Dialog, error) {
+func (establisher *dialogEstablisher) CreateDialog(
+	peerID identity.Identity,
+	peerContact dto_discovery.Contact,
+) (communication.Dialog, error) {
 	var dialog *dialog
 
-	log.Info(establisherLogPrefix, fmt.Sprintf("Connecting to: %#v", contact))
-	contactAddress, err := establisher.contactAddressFactory(contact)
+	log.Info(establisherLogPrefix, fmt.Sprintf("Connecting to: %#v", peerContact))
+	peerAddress, err := establisher.peerAddressFactory(peerContact)
 	if err != nil {
-		return dialog, fmt.Errorf("failed to connect to: %#v. %s", contact, err)
+		return dialog, fmt.Errorf("failed to connect to: %#v. %s", peerContact, err)
 	}
 
-	contactCodec := NewCodecSecured(communication.NewCodecJSON(), establisher.mySigner, identity.NewVerifierSigned())
-	contactSender := nats.NewSender(contactAddress.GetConnection(), contactCodec, contactAddress.GetTopic())
+	peerCodec := establisher.newCodecForPeer(peerID)
 
-	response, err := contactSender.Request(&dialogCreateProducer{
+	peerSender := establisher.newSenderToPeer(peerAddress, peerCodec)
+	err = establisher.negotiateDialog(peerSender)
+	if err != nil {
+		return dialog, err
+	}
+
+	dialog = establisher.newDialogToPeer(peerAddress, peerCodec)
+	log.Info(establisherLogPrefix, fmt.Sprintf("Dialog established with: %#v", peerContact))
+
+	return dialog, nil
+}
+
+func (establisher *dialogEstablisher) negotiateDialog(sender communication.Sender) error {
+	response, err := sender.Request(&dialogCreateProducer{
 		&dialogCreateRequest{
-			IdentityId: establisher.myIdentity.Address,
+			PeerID: establisher.myID.Address,
 		},
 	})
 	if err != nil {
-		return dialog, fmt.Errorf("dialog creation error. %s", err)
+		return fmt.Errorf("dialog creation error. %s", err)
 	}
 	if response.(*dialogCreateResponse).Reason != 200 {
-		return dialog, fmt.Errorf("dialog creation rejected. %#v", response)
+		return fmt.Errorf("dialog creation rejected. %#v", response)
 	}
 
-	dialog = establisher.newDialogToContact(contactAddress, contactCodec)
-	log.Info(establisherLogPrefix, fmt.Sprintf("Dialog established with: %#v", contact))
+	return nil
+}
 
-	return dialog, nil
+func (establisher *dialogEstablisher) newCodecForPeer(peerID identity.Identity) *codecSecured {
+
+	return NewCodecSecured(
+		communication.NewCodecJSON(),
+		establisher.mySigner,
+		identity.NewVerifierIdentity(peerID),
+	)
+}
+
+func (establisher *dialogEstablisher) newSenderToPeer(
+	peerAddress *discovery.AddressNATS,
+	peerCodec *codecSecured,
+) communication.Sender {
+
+	return nats.NewSender(
+		peerAddress.GetConnection(),
+		peerCodec,
+		peerAddress.GetTopic(),
+	)
 }
 
-func (establisher *dialogEstablisher) newDialogToContact(
-	contactAddress *discovery.AddressNATS,
-	contactCodec communication.Codec,
+func (establisher *dialogEstablisher) newDialogToPeer(
```

Comment:
`newDialogForPeer`

Reply:
But I want to send messages to another peer

Reply:
ok, makes sense